File: src/main/java/com/dude/dms/backend/repositories/TagRepository.java
package com.dude.dms.backend.repositories;

import com.dude.dms.backend.data.rules.PlainTextRule;
import com.dude.dms.backend.data.rules.RegexRule;
import com.dude.dms.backend.data.tags.Tag;
import org.springframework.data.jpa.repository.JpaRepository;

import java.util.List;
import java.util.Optional;
import java.util.Set;

public interface TagRepository extends JpaRepository<Tag, Long> {

    Optional<Tag> findByName(String name);

    Set<Tag> findByPlainTextRules(PlainTextRule rule);

    Set<Tag> findByRegexRules(RegexRule rule);

    List<Tag> findTop10ByNameContaining(String name);

    long countByNameContaining(String name);

    List<Tag> findTop10ByNameContainingIgnoreCase(String name);

    long countByNameContainingIgnoreCase(String name);
}

File: src/main/java/com/dude/dms/backend/repositories/TagRepository.kt
package com.dude.dms.backend.repositories

import com.dude.dms.backend.data.Tag
import com.dude.dms.backend.data.docs.Attribute
import com.dude.dms.backend.data.docs.Doc
import com.dude.dms.backend.data.mails.Mail
import com.dude.dms.backend.data.mails.MailFilter
import com.dude.dms.backend.data.rules.PlainTextRule
import com.dude.dms.backend.data.rules.RegexRule
import org.springframework.data.jpa.repository.JpaRepository

interface TagRepository : JpaRepository<Tag, Long> {

    fun findByName(name: String): Tag?

    fun findByPlainTextRules(rule: PlainTextRule): Set<Tag>

    fun findByRegexRules(rule: RegexRule): Set<Tag>

    fun findByMailFilters(mailFilter: MailFilter): Set<Tag>

    fun findTop10ByNameContaining(name: String): Set<Tag>

    fun countByNameContaining(name: String): Long

    fun findTop10ByNameContainingIgnoreCase(name: String): Set<Tag>

    fun countByNameContainingIgnoreCase(name: String): Long

    fun findByDocs(doc: Doc): Set<Tag>

    fun findByMails(mail: Mail): Set<Tag>

    fun findByAttributes(attribute: Attribute): Set<Tag>

    fun countByAttributes(attribute: Attribute): Long
}